### PR TITLE
CoAP implementation

### PR DESCRIPTION
**Pull request by xkevin190**
gcoap basic implementation, currently not functional

### CODE REVIEW

File: sys/include/net/tfcoap.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * @{
+ * @file        tfcoap.h
+ * @brief       Constrained Application Protocol
+ *
+ * @author      Locha Mesh Developers <developers@locha.io>
+ * @}
+ */
+
+#include "net/gcoap.h"
+
```
Comment (jeandudey) [inline sys/include/net/tfcoap.h:29]: add:

```c
#ifdef __cplusplus
extern "C" {
#endif
```

File: sys/include/net/tfcoap.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * @ingroup     net_coap
+ *
+ * @{
+ * @file        tfcoap.h
+ * @brief       Constrained Application Protocol
+ *
+ * @author      Locha Mesh Developers <developers@locha.io>
+ * @}
+ */
+
```
Comment (jeandudey) [inline sys/include/net/tfcoap.h:27]: add:

```c
#ifndef NET_TFCOAP_H
#define NET_TFCOAP_H
```

File: sys/include/net/tfcoap.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * @defgroup    net_coap
+ * @ingroup     net_coap
+ *
+ * @{
+ * @file        tfcoap.h
+ * @brief       Constrained Application Protocol
+ *
+ * @author      Locha Mesh Developers <developers@locha.io>
+ * @}
```
Comment (jeandudey) [inline sys/include/net/tfcoap.h:25]:
```suggestion
```

File: sys/net/coap/tfcoap.c [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+
+ssize_t stats_handler(coap_pkt_t* pdu, uint8_t *buf, size_t len, void *ctx)
+{
+    puts("execute\n");
+    return 0;
+}
+
+void tf_coat_init(void){
+    printf("EXECUTE HEREEEE!!!!!!!!!!!!!!!!!!!!!!!!!!1 \n");
```
Comment (jeandudey) [inline sys/net/coap/tfcoap.c:79]:
```suggestion
    DEBUG("EXECUTE HEREEEE!!!!!!!!!!!!!!!!!!!!!!!!!!1 \n");
```

File: sys/include/net/tfcoap.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+#include "net/gcoap.h"
+
+void tf_coat_init(void);
+
+ssize_t encode_link(const coap_resource_t *resource, char *buf,
+                            size_t maxlen, coap_link_encoder_ctx_t *context);
+
+ssize_t stats_handler(coap_pkt_t* pdu, uint8_t *buf, size_t len, void *ctx);
```
Comment (jeandudey) [inline sys/include/net/tfcoap.h:35]: add:

```c
#ifdef _cplusplus
} /* extern "C" */
#endif

#endif /* NET_TFCOAP_H */
/** @} */
```

**Comment by xkevin190** [comment]
the code needs improvements you will find a little mess, this is going to improve, what interests me is: why does it not work?

**Comment by jeandudey** [comment]
The code seems okay without errors aside from the C++ compat declarations and include guards.

A good protip is to enable the debug on `gcoap` module on RIOT, by setting `ENABLE_DEBUG` to 1 on all coap related files (`RIOT/sys/gcoap/`).

**Comment by xkevin190** [comment]
this has already been done, but without having an answer as to why it doesn't work, try adding the nanocoap server but this affects the functionality of the ap

File: sys/net/coap/tfcoap.c [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        return gcoap_response(pdu, buf, len, COAP_CODE_INTERNAL_SERVER_ERROR);
+    }
+
+    return 0;
+}
+
+void tf_coat_init(void){
+
+    gcoap_register_listener(&listener);
+}
```
Comment (luisan00) [inline sys/net/coap/tfcoap.c:131]: no new line at the end
```suggestion
}

```

File: sys/net/coap/tfcoap.c [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    cJSON_AddNumberToObject(root, "voltage", voltage * 1.0);
+    cJSON_AddNumberToObject(root, "avg_current", avg_current * 1.0);
+    cJSON_AddNumberToObject(root, "avg_power", avg_power * 1.0);
+    cJSON_AddNumberToObject(root, "avg_temp", avg_power * 1.0);
+    cJSON_AddNumberToObject(root, "free_memory", free_memory * 1.0);
+
+    cJSON* ap_root = cJSON_AddObjectToObject(root, "ap");
+    cJSON_AddStringToObject(ap_root, "ssid", ap_ssid);
+
+    // ****  STA is not avaliable in this version  **** //
```
Comment (luisan00) [inline sys/net/coap/tfcoap.c]:
```suggestion
    // ****  STA is not available in this version  **** //
```

**Pull request merged**